full/coupler_main.F90

```diff
@@ -1340,40 +1340,71 @@ subroutine coupler_init
-            slow_ice_ocean_pelist(:) = Ice%slow_pelist(:)
-         else
-            allocate(slow_ice_ocean_pelist(ocean_npes))
-            slow_ice_ocean_pelist(:) = Ocean%pelist(:)
-         endif
+    if (do_atmos) then
+      if (Ice%shared_slow_fast_PEs) then
+        ! Fast and slow ice processes occur on the same PEs.
+        allocate( Ice%pelist  (ice_npes) )
+        Ice%pelist(:) = Ice%fast_pelist(:)
+        allocate( Ice%slow_pelist(ice_npes) )
+        Ice%slow_pelist(:) = Ice%fast_pelist(:)
+        if(concurrent) then
+          allocate(slow_ice_ocean_pelist(ocean_npes+ice_npes))
+          slow_ice_ocean_pelist(1:ice_npes) = Ice%slow_pelist(:)
+          slow_ice_ocean_pelist(ice_npes+1:ice_npes+ocean_npes) = Ocean%pelist(:)
+        else
+          if(ice_npes .GE. ocean_npes) then
+             allocate(slow_ice_ocean_pelist(ice_npes))
+             slow_ice_ocean_pelist(:) = Ice%slow_pelist(:)
+          else
+             allocate(slow_ice_ocean_pelist(ocean_npes))
+             slow_ice_ocean_pelist(:) = Ocean%pelist(:)
+          endif
+        endif
+      else
+        ! Fast ice processes occur a subset of the atmospheric PEs, while
+        ! slow ice processes occur on the ocean PEs.
+        allocate( Ice%slow_pelist(ocean_npes) )
+        Ice%slow_pelist(:) = Ocean%pelist(:)
+        allocate( Ice%pelist  (ice_npes+ocean_npes) )
+        ! Set Ice%pelist() to be the union of Ice%fast_pelist and Ice%slow_pelist.
+        Ice%pelist(1:ice_npes) = Ice%fast_pelist(:)
+        Ice%pelist(ice_npes+1:ice_npes+ocean_npes) = Ocean%pelist(:)
+        allocate(slow_ice_ocean_pelist(ocean_npes))
+        slow_ice_ocean_pelist(:) = Ocean%pelist(:)
+      endif
+    elseif (.not.do_atmos) then
+      ! In the no atmos cases, shared_slow_fast_PEs is not enough to distinguish
+      ! the slow and fast ice procesor layout; slow_ice_with_ocean should be used instead.
+      if (slow_ice_with_ocean) then
+        ! data atmos, using combined ice-ocean driver
+        ! Both fast ice and slow ice processes occur on the same PEs,
+        ! since the Atmos and Ocean PEs are shared
+        allocate( Ice%slow_pelist(ocean_npes) )
+        Ice%slow_pelist(:) = Ocean%pelist(:)
+        allocate( Ice%pelist  (ice_npes) )
+        Ice%pelist(1:ice_npes) = Ice%fast_pelist(:)
+        allocate(slow_ice_ocean_pelist(ocean_npes))
+        slow_ice_ocean_pelist(:) = Ocean%pelist(:)
+      else
+        ! data atmos, not using combined ice-ocean driver
+        allocate( Ice%pelist  (ice_npes) )
+        Ice%pelist(:) = Ice%fast_pelist(:)
+        allocate( Ice%slow_pelist(ice_npes) )
+        Ice%slow_pelist(:) = Ice%fast_pelist(:)
+        if(ice_npes .GE. ocean_npes) then
+           allocate(slow_ice_ocean_pelist(ice_npes))
+           slow_ice_ocean_pelist(:) = Ice%slow_pelist(:)
+        else
+           allocate(slow_ice_ocean_pelist(ocean_npes))
+           slow_ice_ocean_pelist(:) = Ocean%pelist(:)
+        endif
```

Comment:
I'm not going to hold up the approval based on this case, but is this a scenario the ocean group actually runs?